draft-ietf-quic-transport.md

```diff
@@ -1431,63 +1429,78 @@ a single packet.
-were sent in marked IP packets, then the path or peer remove ECN markings. The
-endpoint records this path as not being ECN capable and it ceases marking of
-packets.
 
-IP packets sent on a new path SHOULD be marked with ECT(0) or ECT(1) to verify
-that the new path supports ECN, see {{ecn-connection-migration}}.
+## Using Explicit Congestion Notification {#using-ecn}
 
-### Continuous Verification of ECN {#ecn-continuous-verification}
+QUIC endpoints SHOULD use Explicit Congestion Notification (ECN) {{!RFC3168}} to
+detect and respond to network congestion.  ECN allows a network node to indicate
+congestion in the network by setting a codepoint in the IP header of a packet
+instead of dropping it.  Endpoints react to congestion by reducing their sending
+rate in response, as described in {{QUIC-RECOVERY}}.
 
-If the ECN capability check was successful and the endpoint continues to send
-ECT marked packets then continuous verification is applied. This is to detect
-any cases when ECN field is bleached, that is, zeroed out by a network node,
-likely as the result of a routing changes since the ECN capability check.
+To use ECN, QUIC endpoints must first determine whether a path and peer support
+ECN marking. Verifying the path occurs at the beginning of a connection and when
+the connection migrates to a new path ({{ecn-connection-migration}}).
 
-For each received ACK_ECN frame, the total number of newly acknowledged packets
-can be compared to the total increase in ECN counters. If the increase in ECN
-counters is less, then an ECN failure has occurred and ECN should be disabled.
-ECN is also disabled in case an ACK frame is received acknowledging any ECT sent
-packet.
+Each endpoint independently verifies and enables ECN for the path from it to the
+peer.
+
+To verify that both a path and the peer support ECN, an endpoint MUST set one of
+the ECN Capable Transport (ECT) codepoints -- ECT(0) or ECT(1) -- in the IP
+header {{!RFC8311}} of all outgoing packets.
+
+If an ECT codepoint set in the IP header is not bleached or otherwise corrupted
+by a network device, then a received packet contains either the codepoint sent
+by the peer or the Congestion Experienced (CE) codepoint set by a network device
+that is experiencing congestion.
+
+On receiving a packet with an ECT or CE codepoint, an endpoint that supports ECN
+increases the corresponding local ECT(0), ECT(1), or CE count that it maintains
+locally.
+
+A packet detected by a receiver as a duplicate does not affect the receiver's
+local ECN codepoint counts to mitigate security concerns ({{security-ecn}}).
+
+An endpoint that supports ECN responds to packets received with ECT or CE
+codepoints when necessary (see {{processing-and-ack}}) with an ACK_ECN frame
+carrying the current state of its ECN counters (see {{frame-ack-ecn}}).
+
+If an endpoint receives a packet without an ECT or CE codepoint, it responds per
```

Comment:
So, my proposal using three counters so far is using ACK_ECN frame as long as there are any ECN marked packets in the range being acknowledged. This formulation can be interpreted such that ACK frames shall be used to acknowledge all packets that are unmarked. This is a possible change and would be step in the direction of the two counter proposal as well as making it clearer which packets wasn't marked allowing better sender verification if it like to check things by not marking all packets. Is this your intention here?

Reply:
Yes, this is a change from your proposal, but it is necessary to be able to detect all failure modes correctly. 

There's a change earlier in the text that basically says that an endpoint marks all packets if it wants to do ECN. The idea here is that if an ACK is received at all, the endpoint can simply turn off ECN, since that only happens if the network or the peer don't support ECN.